pkg/distributor/level_detection.go

```diff
@@ -0,0 +1,197 @@
+package distributor
+
+import (
+	"bytes"
+	"strconv"
+	"strings"
+	"unicode"
+	"unsafe"
+
+	"github.com/buger/jsonparser"
+	"github.com/prometheus/prometheus/model/labels"
+	"go.opentelemetry.io/collector/pdata/plog"
+
+	"github.com/grafana/loki/v3/pkg/loghttp/push"
+	"github.com/grafana/loki/v3/pkg/logproto"
+	"github.com/grafana/loki/v3/pkg/logql/log/logfmt"
+	"github.com/grafana/loki/v3/pkg/util/constants"
+)
+
+func (l *LevelDetector) allowedLabelsForLevel() map[string]struct{} {
+	allowedFields := l.validationContext.logLevelFields
+	if len(allowedFields) == 0 {
+		return map[string]struct{}{
+			"level": {}, "LEVEL": {}, "Level": {},
+			"severity": {}, "SEVERITY": {}, "Severity": {},
+			"lvl": {}, "LVL": {}, "Lvl": {},
+		}
+	}
+	allowedFieldsMap := make(map[string]struct{}, len(allowedFields))
+	for _, field := range allowedFields {
+		allowedFieldsMap[field] = struct{}{}
+	}
+	return allowedFieldsMap
```

Comment:
`allowedLabelsForLevel()` can be called quite a few times. I think we should allocate the map only once.

Reply:
fixed